hera_cal/io.py

```diff
@@ -7,26 +7,537 @@
+    spaced in the underlying data_array.
+    '''
+
+    # static list of useful metadata to calculate and save
+    HERAData_metas = ['ants', 'antpos', 'freqs', 'times', 'lsts', 'pols',
+                      'antpairs', 'bls', 'times_by_bl', 'lsts_by_bl']
+
+    def __init__(self, input_data, filetype='uvh5'):
+        '''Instantiate a HERAData object. If the filetype is uvh5, read in and store
+        useful metadata (see get_metadata_dict()), either as object attributes or,
+        if input_data is a list, as dictionaries mapping string paths to metadata.
+
+        Arguments:
+            input_data: string data file path or list of string data file paths
+            filetype: supports 'uvh5' (defualt), 'miriad', 'uvfits'
+        '''
+        # initialize as empty UVData object
+        super(HERAData, self).__init__()
+
+        # parse input_data as filepath(s)
+        if isinstance(input_data, str):
+            self.filepaths = [input_data]
+        elif isinstance(input_data, collections.Iterable):  # List loading
+            if np.all([isinstance(i, str) for i in input_data]):  # List of visibility data paths
+                self.filepaths = list(input_data)
+            else:
+                raise TypeError('If input_data is a list, it must be a list of strings.')
+        else:
+            raise ValueError('input_data must be a string or a list of strings.')
+        for f in self.filepaths:
+            if not os.path.exists(f):
+                raise IOError('Cannot find file ' + f)
+
+        # load metadata from file
+        self.filetype = filetype
+        if self.filetype is 'uvh5':
+            # read all UVData metadata from first file
+            temp_paths = copy.deepcopy(self.filepaths)
+            self.filepaths = self.filepaths[0]
+            self.read(read_data=False)
+            self.filepaths = temp_paths
+
+            if len(self.filepaths) > 1:  # save HERAData_metas in dicts
+                for meta in self.HERAData_metas:
+                    setattr(self, meta, {})
+                for path in self.filepaths:
+                    hc = HERAData(path, filetype='uvh5')
+                    meta_dict = self.get_metadata_dict()
+                    for meta in self.HERAData_metas:
+                        getattr(self, meta)[path] = meta_dict[meta]
+            else:  # save HERAData_metas as attributes
+                self._writers = {}
+                for key, value in self.get_metadata_dict().items():
+                    setattr(self, key, value)
+
+        elif self.filetype in ['miriad', 'uvfits']:
+            for meta in self.HERAData_metas:
+                setattr(self, meta, None)  # no pre-loading of metadata
+        else:
+            raise NotImplementedError('Filetype ' + self.filetype + ' has not been implemented.')
+
+    def reset(self):
+        '''Resets all standard UVData attributes, potentially freeing memory.'''
+        super(HERAData, self).__init__()
```

Comment:
just wondering if you find that this actually does free-up the amount of memory you expect?

Reply:
Not as much or as consistently as I'd like, but yes it does clear up some memory. I tested this with %memit